pkg/controllers/securitypolicy_associated_controller.go

```diff
@@ -0,0 +1,246 @@
+/* Copyright © 2022 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package controllers
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+)
+
+// Take these namespaces started with "kube-" or "vmware-" as system namespace.
+// Events in these namespaces will be ignored. "kube-" is for kubernetes system namespace,
+// and "vmware-" is for vmware system namespace.
+var ignoreNs = []string{"kube-", "vmware-"}
+
+// We should consider the below sceneries:
+// When a new added pod whose port name exists in security policy.
+// When a deleted pod whose port name exists in security policy.
+// When a pod's label is changed.
+// When a namespace's label is changed.
+// In summary, we could roughly think if the port name of security policy exists in the
+// new pod or old pod, we should reconcile the security policy.
+
+type EnqueueRequestForNamespace struct {
+	Client client.Client
+}
+
+func (e *EnqueueRequestForNamespace) Create(_ event.CreateEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace create event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Delete(_ event.DeleteEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace delete event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Generic(_ event.GenericEvent, _ workqueue.RateLimitingInterface) {
+	log.V(1).Info("namespace generic event, do nothing")
+}
+
+func (e *EnqueueRequestForNamespace) Update(updateEvent event.UpdateEvent, l workqueue.RateLimitingInterface) {
+	obj := updateEvent.ObjectNew.(*v1.Namespace)
+
+	podList := &v1.PodList{}
+	err := e.Client.List(context.Background(), podList, client.InNamespace(obj.Name))
+	if err != nil {
+		log.Error(err, "failed to list pod in namespace", "namespace", obj.Name)
+		return
+	}
+
+	shouldReconcile := false
+	for _, pod := range podList.Items {
+		if checkPodHasNamedPort(pod, "update") {
+			shouldReconcile = true
+			break
+		}
+	}
+	if !shouldReconcile {
+		log.Info("no pod in namespace is relevant", "namespace", obj.Namespace, "name", obj.Name)
+		return
+	}
+
+	err = reconcileSecurityPolicy(e.Client, podList.Items, l)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
```

Comment:
Do we need retry for this failure?

Reply:
The error occurs only when listing security policy which rarely happens, I think there is no need to retry. Put the request to queue and the controller will be responsible to retry.